src/unix/os390-syscalls.c

```diff
@@ -308,6 +312,10 @@ int epoll_wait(uv__os390_epoll* lst, struct epoll_event* events,
     events[nevents++] = ev;
   }
 
+  if (msg_fd.revents != 0 && msg_fd.fd != -1)
+    if (i == lst->size)
+      events[nevents - 1].is_msg = 1;
```

Comment:
It looks like `nevents - 1` will evaluate to `-1` when you call this function with `maxevents == 0`.

Reply:
> Okay. Can you add an assert(lst->size > 0); just in case then?

What about adding the following checks (have not compiled) at the top of the function?

```c

#define EP_MAX_PFDS (ULONG_MAX / sizeof(struct pollfd))
#define EP_MAX_EVENTS (INT_MAX / sizeof(struct epoll_event))

...

if (!lst || !lst->items || !events) {
  errno = EFAULT;
  return -1;
}

if (lst->size == 0 || lst->size > EP_MAX_PFDS) {
  errno = EINVAL;
  return -1;
}

if (maxevents <= 0 || maxevents > EP_MAX_EVENTS) {
  errno = EINVAL;
  return -1;
}
```
This could be an overkill but it'll bring this epoll_wait closer to the [expected behavior](https://github.com/torvalds/linux/blob/master/fs/eventpoll.c#L2259) and protect from the edge cases you brought up.

Reply:
> It looks like ```nevents - 1``` will evaluate to ```-1``` when you call this function with ```maxevents == 0```.

If ```maxevents == 0```, ```i``` will be ```0```. If ```i == 0```, we'd reach that line only if ```lst->size == 0```, but if that's the case, we'd leave the function right after the ```poll``` call.

Please see my comment above. I think an explicit error handling would make it much clearer how the function behaves.

Reply:
@bnoordhuis can you please take a look at my comment about the additional checks?

Reply:
I'd probably discriminate between pathological but possible cases and "impossible" cases, i.e., cases that are probably indicative of bugs elsewhere. The former should be handled formally, the latter by asserting or aborting.

It's not entirely clear from your description whether you feel in what category the edge cases I brought up fall in but I'll trust your good judgment if you can add a code comment or two explaining your decisions. :-)

Reply:
Thinking about this some more, the conditions you list look okay to me, except maybe that `lst->size == 0` should be interpreted as "sleep for the requested timeout", because that's what epoll_wait() on linux would do; it's basically `select(0, 0, 0, 0, &timeout)` when no fds are registered.

Reply:
I agree. With that I'll need to make sure that `_SET_FDS_MSGS` is not passed UINT64_MAX when `lst->size` is 0.